src/Build/Collections/PropertyDictionary.cs

```diff
@@ -497,17 +586,31 @@ internal void RemoveProperties(IEnumerable<T> other)
         {
             Dictionary<string, string> dictionary;
 
-            lock (_properties)
+            _lock.EnterReadLock();
+
+            try
             {
                 dictionary = new Dictionary<string, string>(_properties.Count, MSBuildNameIgnoreCaseComparer.Default);
 
-                foreach (T property in this)
+                foreach (T property in _properties.Values)
                 {
                     dictionary[property.Key] = property.EscapedValue;
                 }
             }
+            finally
+            {
+                _lock.ExitReadLock();
+            }
 
             return dictionary;
         }
+
+        /// <summary>
+        /// Finalizer to clean up the <see cref="ReaderWriterLockSlim" />.
+        /// </summary>
+        ~PropertyDictionary()
+        {
+            _lock?.Dispose();
```

Comment:
Is this needed? If the finalizer thread is used to run dispose on the lock, why not just cut out the middleman and let the finalizer thread finalize the lock.

Reply:
Does the slim version even require disposing?

---
In reply to: [285822040](https://github.com/microsoft/msbuild/pull/4374#discussion_r285822040) [](ancestors = 285822040)

Reply:
It's not clear to me whether it does or not. I was aping the [`ReaderWriterLockSlim` doc example](https://docs.microsoft.com/en-us/dotnet/api/system.threading.readerwriterlockslim?view=netframework-4.8#examples). But I see several current examples in CoreFX that don't use dispose it, like [`CodePagesEncodingProvider`](https://github.com/dotnet/corefx/blob/5e44821bd670530cd2bf1dd330079f6e2d8e5284/src/System.Text.Encoding.CodePages/src/System/Text/CodePagesEncodingProvider.cs#L14) and [`Privilege`](https://github.com/dotnet/corefx/blob/5e44821bd670530cd2bf1dd330079f6e2d8e5284/src/System.Security.AccessControl/src/System/Security/AccessControl/Privilege.cs#L41). So I guess it's ok to skip it.

Reply:
@stephentoub any reason we're inconsistent?

Reply:
> any reason we're inconsistent?

In Privilege, it's static.

And CodePagesEncodingProvider exposes a static singleton that contains the instance, so it's essentially static there as well.

Reply:
> Does the slim version even require disposing?

ReaderWriterLockSlim wraps several event wait handles.  If the RWLS isn't disposed, they'll be left for finalization.